packages/caravan-health/src/privacy.ts

```diff
@@ -0,0 +1,204 @@
+
+/*
+The p_score is calculated by evaluating the likelihood of self-payments, the involvement of 
+change outputs and the type of transaction based on number of inputs and outputs.
+
+We have 5 categories of transaction type
+- Sweep Spend
+- Simple Spend
+- UTXO Fragmentation
+- Consolidation
+- CoinJoin
+*/
+export function privacyScoreOnIO(transaction: any, client: BlockchainClient): number {
+  const numberOfInputs: number = transaction.vin.length;
+  const numberOfOutputs: number = transaction.vout.length;
+
+  let score: number;
+
+  if (numberOfInputs === 1) {
+    if (numberOfOutputs === 1) {
+      // Sweep Spend (No change Output)
+      // #Input = 1, #Output = 1
+      score = 1 / 2;
+    } else if (numberOfOutputs === 2) {
+      // Simple Spend (Single change output)
+      // #Input = 1, #Output = 2
+      score = 4 / 9;
+    } else {
+      // UTXO Fragmentation
+      // #Input = 1, #Output > 2
+      score = 2 / 3 - 1 / numberOfOutputs;
+    }
+    if (isSelfPayment(transaction,client)) {
+      return score * DENIABILITY_FACTOR;
+    }
+  } else {
+    if (numberOfOutputs === 1) {
+      // Consolidation
+      // #Input >= 2, #Output = 1
+      score = 1 / numberOfInputs;
+
+      // No D.F for consolidation
+    } else {
+      // Mixing or CoinJoin
+      // #Input >= 2, #Output >= 2
+      let x = Math.pow(numberOfOutputs, 2) / numberOfInputs;
+      score = (0.75 * x) / (1 + x);
+      if (isSelfPayment(transaction,client)) {
+        return score * DENIABILITY_FACTOR;
+      }
+    }
+  }
+  return score;
+}
+
+/*
+Determine whether a Bitcoin transaction is a self-payment, meaning the sender and recipient
+are the same entity. To check this, we make an RPC call to the watcher wallet asking for the
+amount that a given address holds. If the call returns a number then it is part of wallet
+otherwise it is not a self payment.
+*/
+function isSelfPayment(transaction: any, client: BlockchainClient): boolean {
+  transaction.vout.forEach(async op => {
+     if(await client.getAddressStatus(op.scriptPubKey.address) === undefined){
+       return false;
+     }
+  })
+  return true;
+}
+
+/* TODO : replace any type to custom types for Transactions and UTXOs*/
+/* 
+In order to score for address reuse we can check the amount being hold by reused UTXOs 
+with respect to the total amount
+*/
+export function addressReuseFactor(utxos: Array<any>): number {
+  let reused_amount : number = 0;
+  let total_amount : number = 0;
+  utxos.forEach((utxo) => {
+    if (utxo.reused) {
+      reused_amount += utxo.amount;
+    }
+    total_amount += utxo.amount;
+  });
+  return reused_amount / total_amount;
+}
+
+/*
+If we are making payment to other wallet types then the privacy score should decrease because 
+the change received will be at address of our wallet type and it will lead to derivation that 
+we still own that amount.
+*/
+export function addressTypeFactor(transactions : Array<any>, walletAddressType : string): number {
```

Comment:
we have explicit script types exported from either @caravan/wallets or @caravan/bitcoin. You should be able to use that here and possibly for some of the matching. I believe there are also some utilities that exist there for determining address type.

Reply:
You also have to make sure that these can be network agnostic (which those functions should help with)